### PR TITLE
Improve overriden target detection to let it handle stackable-traits-style overrides

### PR DESCRIPTION
**Pull request by lihaoyi**
Fixes https://github.com/com-lihaoyi/mill/issues/554

This PR swaps out our detection of overriden targets and commands: 

- Previously, we tried to count how many overriden methods there are upstream of a particular target, statically at its definition site. Any target that has the max number of overriden methods is considered the "final" one, while others are considered overriden. This fails for stackable traits, e.g. the example in the test suite below, as multiple traits can have the same number of static definiton-site overrides but when stacked together on a module one of them ends up winning

- Now, we simply do a reverse walk of the targets list after they have been topo sorted. Since the directionality of overrides is exactly opposite that of the topo sort, we can assume that the first time we see a target of a particular label in the reverse walk it's the "final" one, while any subsequent targets we see with the same label must be the overriden versions

The new mechanism is simpler, and possibly more correct. Hopefully CI would catch any issues...

Added a test case that fails on master and passes on this PR. Also verified manually that the issue reported above can be reproduced on master, and does not reproduce (i.e. it behaves correctly) on this PR.

### CODE REVIEW

**Comment by lefou** [comment]
I love that we get rid of the `overrides` parameter! I always avoided to understand it fully, haha.

**Comment by lihaoyi** [comment]
@lefou looks like the bootstrap tests are failing due to some external plugins. Not sure what your SOP for dealing with these is, but IMO for plugins Mill's own build depends on we should just keep them in the mill repo's contrib folder to simplify this sort of change

**Comment by lefou** [comment]
Bootstrap tests are special and can break when we change some API of Mill. Although you see it fail because of plugins, this issue can also happen when we don't depend on external plugins, but instead use some API that has changed directly. (I remember we had that issue in the past, e.g. when we changed some publishing API.) Our CI now applies a patch file before it runs bootstrap tests, it's `ci/mill-bootstrap.patch` and currently empty. That way, we can customize the `build.sc` or other files before we bootstrap.

I don't think we should deny ourselves to use external plugins, just because we want to run some bootstrap tests. Instead, we can adapt the build (and ad-hoc disable plugins) to run the bootstrap tests.

**Comment by lefou** [comment]
I can prepare a patch file and push it to your branch later

File: main/core/src/mill/eval/Evaluator.scala [hunk 1]
```diff
@@ -94,6 +94,13 @@ case class Evaluator(
       testReporter: TestReporter = DummyTestReporter
   ): Evaluator.Results = {
     val (sortedGroups, transitive) = Evaluator.plan(rootModule, goals)
+    val terminalNames = sortedGroups
+      .keys()
+      .map{
+        case Left(x) => x
+        case Right(x) => (System.identityHashCode(x), x.task.ctx.lineNum, x.segments.render)
+      }
+      .toList
```
Comment (lefou) [inline main/core/src/mill/eval/Evaluator.scala:103]: What happens here and where is `terminalNames` used?

File: main/core/src/mill/eval/Evaluator.scala [hunk 2]
```diff
@@ -659,44 +666,27 @@ object Evaluator {
   def plan(rootModule: BaseModule, goals: Agg[Task[_]]) = {
```
Comment (lefou) [inline main/core/src/mill/eval/Evaluator.scala:666]: Looks like we no longer need that rootModule parameter here?

File: main/core/src/mill/eval/Evaluator.scala (hunk 1 above)
Comment (lihaoyi) [reply main/core/src/mill/eval/Evaluator.scala:103]: It's not used, this is leftover debugging logic. Will remove

File: main/core/src/mill/eval/Evaluator.scala (hunk 2 above)
Comment (lihaoyi) [reply main/core/src/mill/eval/Evaluator.scala:666]: Good catch, seems like we used to use it to try and resolve the "final" implementation of a task to do override count comparisons. Now we are simply finding it via the topo sort, it's no longer necessary

**Review by lefou — APPROVED** [review]
LGTM!

**Pull request merged**